src/index.js

```diff
@@ -30,14 +30,29 @@ const defaultPermissions = 0o666;
 
 const relocateLoader = eval('require(__dirname + "/loaders/relocate-loader.js")');
 
+function hasTypeModule (path) {
+  let root = pathResolve('/');
+  while ((path = pathResolve(path, '..')) !== root) {
+    try {
+      return JSON.parse(fs.readFileSync(eval('pathResolve')(path, 'package.json')).toString()).type === 'module';
+    }
+    catch (e) {
+      if (e.code === 'ENOENT')
+        continue;
+      throw e;
+    }
+  }
+}
+
 module.exports = ncc;
 function ncc (
   entry,
   {
     cache,
     customEmit = undefined,
+    esm = entry.endsWith('.mjs') || !entry.endsWith('.cjs') && hasTypeModule(entry),
     externals = [],
-    filename = 'index' + (entry.endsWith('.cjs') ? '.cjs' : '.js'),
+    filename = 'index' + (!esm && entry.endsWith('.cjs') ? '.cjs' : !esm || hasTypeModule(entry) ? '.js' : '.mjs'),
```

Comment:
I'm curious why this doesn't follow the same logic as the CLI above on line 233:

```js
const ext = buildFile.endsWith('.cjs') ? '.cjs' : buildFile.endsWith('.mjs') ? '.mjs' : '.js';
```

Seems like we should preserve the file extension here too

Reply:
Good point, I've cleaned the logic up a bit let me know how that looks.